platform/pulpcore/download/single.py

```diff
@@ -0,0 +1,291 @@
+from gettext import gettext as _
+from logging import getLogger
+from threading import RLock
+
+
+from .delegation import delegate
+from .error import DownloadFailed
+from .validation import ValidationError
+
+
+log = getLogger(__name__)
+
+
+class Download:
+    """
+    An ABSTRACT download object.
+    Represents a unit of work to download a file. A download is a
+    command-pattern object.
+
+    Attributes:
+        url (str): A file download URL.
+        writer (Writer): An object used to store downloaded file.
+        validations (list): Collection of Validations to be applied.
+        context (Context): The download context.
+        retries (int): The total number of retries possible.
+        delegate (Delegate): A download delegate.
+        attachment: Arbitrary object attached to the download.
+        reply: The remote server reply.
+
+    Any method decorated with `@delegate` may be forwarded to the `delegate` for
+    implementation.  The delegate must define a public method with the same name
+    (without the `_` prefix).  For example, for `_on_error()`, the delegate would
+    define a method named `on_error(self, download, error)`.  Note: The `download`
+    is injected (passed) as the first argument.
+
+    Notes:
+        The validations are applied in order listed.
+        Settings may be shared between downloads.
+    """
+
+    # Data transfer buffer size in bytes.
+    BLOCK = 102400
+    # Number of retires on failed download.
+    RETRIES = 1
+
+    __slots__ = (
+        'url',
+        'writer',
+        'validations',
+        'context',
+        'retries',
+        'delegate',
+        'attachment',
+        'reply'
+    )
+
+    def __init__(self, url, writer):
+        """
+        Args:
+            url (str): A file download URL.
+            writer (Writer): An object used to store downloaded file.
+        """
+        super(Download, self).__init__()
+        self.url = url
+        self.writer = writer
+        self.validations = []
+        self.context = Context()
+        self.retries = Download.RETRIES
+        self.delegate = None
+        self.attachment = None
+        self.reply = None
+
+    def _prepare(self):
+        """
+        Prepare to be executed.
+        """
+        log.debug(_('Prepare: %(d)s'), {'d': self})
```

Comment:
Agreed.  Unfortunately, logging does not support new-style string formatting.  The string should only be formatted when the log level is sufficient for the message to be logged.  In other words, we don't want to format the string when the log level is not at DEBUG.